Move Swift remote state to backend

fatmcgav opened the pull request:
Move the existing Swift remote state to an official backend. 

This will enable additional functionality to be added, such as support for locking and environments. 

TODO: 

- [x] Add tests
- [x] Add documentation

fatmcgav commented:
@jtopjian Would appreciate a quick review on this one whilst I'm still working on tests :)

jtopjian commented:
@fatmcgav I'm not familiar with how the new backend system works, but everything looks good from a cursory review. Nice work. :)

I'll poke at it in more detail once there are some tests in place.

fatmcgav commented:
@jtopjian I've just added some acceptance tests, which all pass for me... 

>make testacc TEST=./backend/remote-state/swift
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/05/07 22:07:52 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./backend/remote-state/swift -v  -timeout 120m
=== RUN   TestBackend_impl
--- PASS: TestBackend_impl (0.00s)
=== RUN   TestBackendConfig
--- PASS: TestBackendConfig (0.58s)
=== RUN   TestBackend
--- PASS: TestBackend (1.66s)
        backend_test.go:22: swift backend acceptance tests enabled
        testing.go:58: TestBackend: named states not supported in *swift.Backend, skipping
=== RUN   TestRemoteClient_impl
--- PASS: TestRemoteClient_impl (0.00s)
=== RUN   TestRemoteClient
2017/05/07 22:08:08 [DEBUG] New state was assigned lineage "48e0c5fa-1c96-46ac-be1b-6a8932c86ba6"
2017/05/07 22:08:08 [TRACE] Preserving existing state lineage "48e0c5fa-1c96-46ac-be1b-6a8932c86ba6"
2017/05/07 22:08:08 [DEBUG] Got to ensureContainerExists...
2017/05/07 22:08:08 [DEBUG] Creating container terraform-state-swift-test-590f8cb8
2017/05/07 22:08:09 [DEBUG] Putting object tfstate.tf at path terraform-state-swift-test-590f8cb8
2017/05/07 22:08:09 [DEBUG] Getting object tfstate.tf at path terraform-state-swift-test-590f8cb8
2017/05/07 22:08:09 [DEBUG] Deleting object tfstate.tf at path terraform-state-swift-test-590f8cb8
2017/05/07 22:08:09 [DEBUG] Getting object tfstate.tf at path terraform-state-swift-test-590f8cb8
2017/05/07 22:08:10 [DEBUG] Object doesn't exist to download.
--- PASS: TestRemoteClient (1.91s)
        backend_test.go:22: swift backend acceptance tests enabled
PASS
ok      github.com/hashicorp/terraform/backend/remote-state/swift       4.204s

I need to extend the tests to cover archive and expiry options, but I'm not sure on best method to do that atm, as the default `remote.TestClient` doesn't cater for any 'special' functionality. 

Let me know your thoughts on changes... :) 

Gav

jtopjian commented on backend/remote-state/swift/backend_test.go:
```diff
@@ -0,0 +1,84 @@
+		"path": path,
+	}).(*Backend)
+
+	// createSwiftContainer(t, b.client, path)
```
Is this no longer used? Can the above line and the `createSwiftContainer` function be removed?

fatmcgav replied:
Yeh, `createSwiftContainer` is no longer required, as the container gets created by the backend [here](https://github.com/hashicorp/terraform/pull/13871/files#diff-9387be13e5c7f0bf51122460f857fc1fR84)

Will remove :)

jtopjian replied:
Sounds good. Might want to give me another hour or two to finish a review. I'm at a conference and making my way through the code when I can :)

fatmcgav replied:
No worries... I'm guna take a stab at the additional tests for `archive_path` and `expire_after` aswell...

jtopjian commented on backend/remote-state/swift/client.go:
```diff
@@ -0,0 +1,116 @@
+}
+
+func (c *RemoteClient) ensureContainerExists() error {
+	log.Println("[DEBUG] Got to ensureContainerExists...")
```
Can probably nix this.

fatmcgav replied:
Yeh, the log lines were useful to a point... Agree that line is superfluous though...

jtopjian commented:
@fatmcgav with regard to client testing, check out the `consul/client.go` and `consul/client_test.go`, notably with the `Gzip` option. Could testing expire and archive be done similarly?

fatmcgav commented:
@jtopjian I'm currently trying out something similar to [`s3/backend_test.go#TestBackendExtraPaths `](https://github.com/hashicorp/terraform/blob/master/backend/remote-state/s3/backend_test.go#L115-L227)

fatmcgav commented:
@jbardin / @mitchellh 
Apologies for the direct mention, however thought it was probably the easiest way of getting assistance :) 

I'm trying to work out how to test the `archive` functionality, and I've come up with the following test flow:

1. Create a `NewState()` object.
2. Write and Persist state.
3. Update state.
4. Write and Persist state.
5. Check for archive container and archive objects.

The bit I'm struggling with is step 3 above, as I can't seem to find an easy method of incrementing the state... I've been looking through the `Consul` and `S3` backends, but can't seem them doing anything similar...  

The current Test code I've got is here: https://gist.github.com/fatmcgav/5137ed266426c14f2433dd6b310a8abd

Any pointers greatly appreciated.

fatmcgav commented:
@jtopjian I've just pushed a few more commits. Most interesting is probably b573d5b which adds `TestBackendArchive`. 

It works for me locally, but would be useful to confirm it works for someone else :) 
Probably needs a bit of a clean up to remove some of the extra log lines, but other than that, I think it's pretty much there... 

With regards to testing the `expire_after` option, I can do basically the same thing, however it's going to have to involve sticking some sleeps into the test so that I can validate that objects are actually being expired. 
Or just take that functionality as granted... 

Thoughts?

jtopjian commented:
Cool! I'll test this out within the next few days.

With regard to sleep, I think if you can keep it within 30 seconds, that should be fine.

jtopjian commented:
@fatmcgav This checks out for me, too. Nice work :)

Let me know when this is ready for a final review. No rush, so if you want to try making a test for Expire, go for it. I'll also see if someone else can review the backend implementation specifically.

fatmcgav commented:
Cool, good to hear... 

Yeh, I'll try and knock up a quick `expire_after` test and then it's GTM :)

fatmcgav commented:
@jtopjian I've looped back around to this one, and whilst looking at it again, I had a thought that this is probably a good time to deprecate the `path` and `archive_path` config variables, in favour of `container` and `archive_container`, which I think is a more accurate description of the purpose. 

For time-being, I've added support for falling back to `path` and `archive_path` if the new variables aren't provided, which we could then remove in 0.11... 
I've also added a test for using the `path` variable, however that's failing in [testing.go:26](https://github.com/hashicorp/terraform/blob/6a3a3b3b05a8d828b09d1fd9aabecb19031234e1/backend/testing.go#L25-L27) due to a 'deprecation' warning being thrown. Failing on a warning feels a bit overzealous to me, or at-least failing on a deprecation warning is... 

Thoughts?

```
make testacc TEST=./backend/remote-state/swift
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/06/06 09:13:32 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./backend/remote-state/swift -v  -timeout 120m
=== RUN   TestBackend_impl
--- PASS: TestBackend_impl (0.00s)
=== RUN   TestBackendConfig
2017/06/06 09:13:42 [DEBUG] Archive_container set, enabling object versioning
--- PASS: TestBackendConfig (1.63s)
        backend_test.go:26: swift backend acceptance tests enabled
=== RUN   TestBackend
--- PASS: TestBackend (1.33s)
        backend_test.go:26: swift backend acceptance tests enabled
        testing.go:58: TestBackend: named states not supported in *swift.Backend, skipping
=== RUN   TestBackendPath
--- FAIL: TestBackendPath (0.00s)
        backend_test.go:26: swift backend acceptance tests enabled
        backend_test.go:77: [DEBUG] Generating backend config
        testing.go:26: warnings: ["path": [DEPRECATED] Use container instead]
=== RUN   TestBackendArchive
2017/06/06 09:13:44 [DEBUG] Archive_container set, enabling object versioning
2017/06/06 09:13:44 [DEBUG] New state was assigned lineage "a40ef259-8cf6-43c7-b129-7cbc134267ff"
2017/06/06 09:13:44 [TRACE] Preserving existing state lineage "a40ef259-8cf6-43c7-b129-7cbc134267ff"
2017/06/06 09:13:44 [DEBUG] Creating archive container terraform-state-swift-test-59366437_archive
2017/06/06 09:13:44 [DEBUG] Enabling Versioning on container terraform-state-swift-test-59366437
2017/06/06 09:13:44 [DEBUG] Creating container terraform-state-swift-test-59366437
2017/06/06 09:13:44 [DEBUG] Putting object tfstate.tf in container terraform-state-swift-test-59366437
2017/06/06 09:13:45 [DEBUG] Getting object tfstate.tf in container terraform-state-swift-test-59366437
2017/06/06 09:13:45 [TRACE] Preserving existing state lineage "a40ef259-8cf6-43c7-b129-7cbc134267ff"
2017/06/06 09:13:45 [TRACE] Preserving existing state lineage "a40ef259-8cf6-43c7-b129-7cbc134267ff"
2017/06/06 09:13:45 [TRACE] Preserving existing state lineage "a40ef259-8cf6-43c7-b129-7cbc134267ff"
2017/06/06 09:13:45 [DEBUG] Creating archive container terraform-state-swift-test-59366437_archive
2017/06/06 09:13:45 [DEBUG] Enabling Versioning on container terraform-state-swift-test-59366437
2017/06/06 09:13:45 [DEBUG] Creating container terraform-state-swift-test-59366437
2017/06/06 09:13:45 [DEBUG] Putting object tfstate.tf in container terraform-state-swift-test-59366437
2017/06/06 09:13:46 [TRACE] Preserving existing state lineage "a40ef259-8cf6-43c7-b129-7cbc134267ff"
2017/06/06 09:13:46 [TRACE] Preserving existing state lineage "a40ef259-8cf6-43c7-b129-7cbc134267ff"
--- PASS: TestBackendArchive (4.91s)
        backend_test.go:26: swift backend acceptance tests enabled
        backend_test.go:106: state1 lineage = a40ef259-8cf6-43c7-b129-7cbc134267ff, serial = 0
        backend_test.go:143: archiveObjects len = 1. Contents = [00atfstate.tf/1496736825.59697]
        backend_test.go:165: Attempting to download object 00atfstate.tf/1496736825.59697 from container terraform-state-swift-test-59366437_archive
        backend_test.go:150: Archive data downloaded... Looks like: &{body:0xc4201e4500 mu:{state:0 sema:0} closed:false rerr:<nil> fn:0x12ff850 earlyCloseFn:0x12ff7e0}
        backend_test.go:156: Archive state lineage = a40ef259-8cf6-43c7-b129-7cbc134267ff, serial = 0, lineage match = true
=== RUN   TestRemoteClient_impl
--- PASS: TestRemoteClient_impl (0.00s)
=== RUN   TestRemoteClient
2017/06/06 09:13:49 [DEBUG] New state was assigned lineage "ea3359e1-2c96-463f-ad9e-35b7672afe10"
2017/06/06 09:13:49 [TRACE] Preserving existing state lineage "ea3359e1-2c96-463f-ad9e-35b7672afe10"
2017/06/06 09:13:49 [DEBUG] Creating container terraform-state-swift-test-5936643c
2017/06/06 09:13:49 [DEBUG] Putting object tfstate.tf in container terraform-state-swift-test-5936643c
2017/06/06 09:13:49 [DEBUG] Getting object tfstate.tf in container terraform-state-swift-test-5936643c
2017/06/06 09:13:49 [DEBUG] Deleting object tfstate.tf in container terraform-state-swift-test-5936643c
2017/06/06 09:13:50 [DEBUG] Getting object tfstate.tf in container terraform-state-swift-test-5936643c
2017/06/06 09:13:50 [DEBUG] Object doesn't exist to download.
--- PASS: TestRemoteClient (2.29s)
        backend_test.go:26: swift backend acceptance tests enabled
FAIL
exit status 1
FAIL    github.com/hashicorp/terraform/backend/remote-state/swift       10.194s
make: *** [testacc] Error 1
```

jtopjian commented:
@fatmcgav Just skimming this, I think there are two ways to go:

1. Either catch the error and recover.
2. Don't test with deprecated options. 

Question about path vs container: Did path always mean "container" or could you do "container/pseudo-directory"? If the latter works, I think "path" might still be the more intuitive name. But if this only works when specifying a container, then definitely change the name :)

fatmcgav commented:
@jtopjian I did try to find a way to catch the Warning and continue, but couldn't see anything obvious in existing examples...
@stack72 Are you able to suggest a solution?

WRT `path` vs `container`, I'd assumed that `path` was always equal to the `container`. However now you mention 'pseudo-directories', it sounds like that's actually a thing... https://docs.openstack.org/user-guide/cli-swift-pseudo-hierarchical-folders-directories.html

But is that something that we want/need to support?

jtopjian commented:
> But is that something that we want/need to support?

It's a tricky situation. Personally, I'd never use a pseudo-container. I'd just use a plain object and call it a day. I'm sure most people would do the same. However, everyone is different, so it's best to always be mindful of that.

Without delving into a bike shed, what about `container_path` ?

fatmcgav commented:
So I had a look, and it looks like folders don't currently work anyway... 

I think it would be a sensible future improvement to support folders, probably by enabling the object name to be specified. For the time being will just document the limitation...

jtopjian commented:
I think that's a good way to go. IMO, I would stick with just specifying a container name, document that pseudo-directories/containers/folders aren't currently supported, and leave it like that until there's enough support to warrant pseudo-directories.

fatmcgav commented:
OK, I've just pushed 60396fc which adds a note regarding the support of pseduo folders. 

I've also improved the test for the `path` config, so hopefully that should round this out... 

Just need to squash them down now... :)

jtopjian commented:
@fatmcgav I really appreciate all of the work you've put into this. Unfortunately this will have to wait until the next release. I'd like someone else to review the remote stuff just to confirm it's good. Hang in there :)

fatmcgav commented:
@jtopjian No worries, makes sense for a 2nd or 3rd pair of eyes on it :)

fatmcgav commented:
@jtopjian This one's going to get interesting to merge now, seeing as the Openstack provider has been pulled out... 

I'll see if I can rework it for 0.10...

fatmcgav commented:
OK, so I've done a quick re-work to support the rename. Need to run acceptance tests, but thought I'd just push this quickly...

fatmcgav commented:
@apparentlymart / @jbardin Is the intention for `backends` to remain part of TF core? 

Also it feels a bit... icky... referencing a `terraform-providers` repo to support the backend. However I note the same change has been applied to the `S3` backend...

jbardin commented:
Hi @fatmcgav,

There's currently no plan to split backends out of the core codebase. 
I agree the situation of importing the provider for the backend seems less than ideal, but as you note, we already have to do the same thing for S3.

While I haven't looked closely at this PR, I assume the import is for the same reason as S3; to share the existing config and initialization code with the provider. After 0.10 I was going to take a look into breaking shared pieces out of the AWS provider into their own package (in the same repo) to see if it offers any architectural advantage. If it seems worthwhile we could do the same for the openstack provider.

fatmcgav commented:
@jbardin 

Ok, cheers for the confirmation. 

And yeh, the import of the `Openstack` package is to enable shared config and init code :)

fatmcgav commented:
@jbardin One other thing then... Is there an easy way of re-running the Travis check? As https://github.com/terraform-providers/terraform-provider-openstack/pull/3 has been merged now, so the tests should pass... :)

jbardin commented:
@fatmcgav,

The provider has the change, but it hasn't been pulled into this repo's vendors yet.

fatmcgav commented:
@jbardin Ah, hadn't spotted that terraform-providers-openstack was vendored... 

Just pushed a vendor update... 

Lets see if it passes :)

jbardin commented:
@fatmcgav,

Thanks for the work to move this over! It looks good to me -- just forgot to mention that we need to remove the dead code in `state/remote`. I'll leave the 👍  review, but wait for that to merge.

fatmcgav commented:
@jbardin Cheers for the review... 

Looks like the deletion of `state/remote/swift` got lost on a rebase... Just pushed an additional commit to remove :)

jtopjian commented:
@fatmcgav Nice work!

fatmcgav commented:
@jbardin Cheers for the merge :) 

@jtopjian Now that's merged, I can take a look at some value add features, such as 'pseudo-locking' and env state...
